src/PythonQtMethodInfo.cpp

```diff
@@ -659,3 +652,78 @@ bool PythonQtSlotInfo::getGlobalShouldAllowThreads()
 {
   return _globalShouldAllowThreads;
 }
+
+
+//! The delegate instance is statically constructed in the UI main thread.
+static PythonQtMetaCallDelegate MetaCallDelegate;
+static std::exception_ptr MetacallException = nullptr;
+
+/**
+ * Delegates the meta call to the PythonQtMetaCallDelegate instance to ensure
+ * that the slot is called in the main UI thread.
+ * The advantage of this method is, that it does rely on private Qt API. The
+ * slot is called in the thread of the MetaCallDelegate object  - that means
+ * in the main UI thread. This allows a Python interpreter running in a
+ * Qt worker thread to call functions of Qt widgets in the main UI thread.
+ */
+static void delegateMetaCall(QObject* obj, PythonQtSlotInfo* slot, void** args)
+{
+  MetacallException = nullptr;
+  MetaCallDelegate.metaObject()->invokeMethod(&MetaCallDelegate, "qt_metacall",
+    Qt::BlockingQueuedConnection,
+    Q_ARG(QObject*, obj),  Q_ARG(void*, slot), Q_ARG(void*, args));
+
+  if (MetacallException)
+  {
+    std::rethrow_exception(MetacallException);
+  }
+}
+
+
+/**
+ * If the current thread is not the main UI thread, then the metacall is
+ * delegated to the MetaCallDelegate. This helper object has been created in
+ * the main UI thread and will execute the metacall in the main UI thread.
+ */
+void PythonQtSlotInfo::invokeQtMethod(QObject* obj, PythonQtSlotInfo* slot, void** args)
+{
+  if (QThread::currentThread() != MetaCallDelegate.thread())
+  {
+    delegateMetaCall(obj, slot, args);
```

Comment:
This assumes that this will only ever be done on objects in the main thread, without a check that the targeted object lives in the main thread. This could break (or at least slow down) a lot of code.